File: R/annotate.R
#' Annotate Biocrates metabolites with cross-reference database IDs
#'
#' Maps Biocrates `short_name`s to the external-database identifiers held in
#' the built-in metabolite dictionary (KEGG, HMDB, PubChem, ChEBI, RefMet,
#' LIPID MAPS, LION). One representative id per metabolite per source; see
#' DESIGN.md §9. The result always has one row per input element, in input
#' order, with `NA` for unmatched names or missing ids — ready to feed into
#' KEGG-map overlays (`pathview`) or external pathway tools.
#'
#' This exposes the id annotations the package already curates so callers no
#' longer need to reach into the internal dictionary directly.
#'
#' @param x Character vector of metabolite `short_name`s, e.g. the
#'   `colnames` of a metabolite assay.
#' @return A data.frame with one row per element of `x` and columns
#'   `metabolite` (the input name), `long_name`, `class`, and the id columns
#'   `kegg_id`, `hmdb_id`, `pubchem_id`, `chebi_id`, `refmet_id`,
#'   `lipidmaps_id`, `lion_id`. Cells are `NA` where unknown.
#' @examples
#' annotate_metabolites(c("Nicotine", "TMAO", "not_a_metabolite"))
#' @export
annotate_metabolites <- function(x) {
  if (!is.character(x)) {
    stop("`x` must be a character vector of metabolite short_names.",
         call. = FALSE)
  }
  ann_cols <- c("long_name", "class", "kegg_id", "hmdb_id", "pubchem_id",
                "chebi_id", "refmet_id", "lipidmaps_id", "lion_id")
  idx <- match(x, metabolite_dict$short_name)
  out <- data.frame(metabolite = x, stringsAsFactors = FALSE)
  for (col in ann_cols) {
    out[[col]] <- metabolite_dict[[col]][idx]
  }
  rownames(out) <- NULL
  out
}

File: python/metabosetr/annotate.py
"""Cross-reference ID annotation (port of ``annotate.R``).

Maps Biocrates ``short_name``s to the external-DB IDs in
:func:`metabolite_dict` (DESIGN.md §6.5 / §9). Always one row per input
element, in input order, with ``None`` for unmatched names or missing IDs --
ready to key a ``pathview`` KEGG-map overlay.
"""

from __future__ import annotations

import pandas as pd

from .reference import metabolite_dict

_ANN_COLS = [
    "long_name",
    "class",
    "kegg_id",
    "hmdb_id",
    "pubchem_id",
    "chebi_id",
    "refmet_id",
    "lipidmaps_id",
    "lion_id",
]


def annotate_metabolites(x) -> pd.DataFrame:
    """Annotate metabolite ``short_name``s with cross-reference IDs.

    ``x`` is a sequence of metabolite short names. Returns a DataFrame with
    columns ``metabolite`` (the input), ``long_name``, ``class``, and the ID
    columns; ``None`` where unknown.
    """
    x = list(x)
    md = metabolite_dict().set_index("short_name")
    out = pd.DataFrame({"metabolite": x})
    for col in _ANN_COLS:
        lookup = md[col].to_dict()  # short_name -> id (or None)
        values = [lookup.get(name) for name in x]
        # normalise residual NaN to None, and assign as an object Series so
        # pandas does not coerce None back to NaN on column assignment.
        values = [
            None if (v is None or (isinstance(v, float) and pd.isna(v))) else v
            for v in values
        ]
        out[col] = pd.Series(values, index=out.index, dtype=object)
    return out
